Use tokio locks in VirtualFile and turn with_file into macro

arpad-m opened the pull request:
## Problem

For #4743, we want to convert everything up to the actual I/O operations of `VirtualFile` to `async fn`.

## Summary of changes

This PR is the last change in a series of changes to `VirtualFile`: #5189, #5190, #5195, #5203, and #5224.

It does the last preparations before the I/O operations are actually made async. We are doing the following things:

* First, we change the locks for the file descriptor cache to tokio's locks that support Send. This is important when one wants to hold locks across await points (which we want to do), otherwise the Future won't be Send. Also, one shouldn't generally block in async code as executors don't like that.
* Due to the lock change, we now take an approach for the `VirtualFile` destructors similar to the one proposed by #5122 for the page cache, to use `try_write`. Similarly to the situation in the linked PR, one can make an argument that if we are in the destructor and the slot has not been reused yet, we are the only user accessing the slot due to owning the lock mutably. It is still possible that we are not obtaining the lock, but the only cause for that is the clock algorithm touching the slot, which should be quite an unlikely occurence. For the instance of `try_write` failing, we spawn an async task to destroy the lock. As just argued however, most of the time the code path where we spawn the task should not be visited. 
* Lastly, we split `with_file` into a macro part, and a function part that contains most of the logic. The function part returns a lock object, that the macro uses. The macro exists to perform the operation in a more compact fashion, saving code from putting the lock into a variable and then doing the operation while measuring the time to run it. We take the locks approach because Rust has no support for async closures. One can make normal closures return a future, but that approach gets into lifetime issues the moment you want to pass data to these closures via parameters that has a lifetime (captures work). For details, see [this](https://smallcultfollowing.com/babysteps/blog/2023/03/29/thoughts-on-async-closures/) and [this](https://users.rust-lang.org/t/function-that-takes-an-async-closure/61663) link. In #5224, we ran into a similar problem with the `test_files` function, and we ended up passing the path and the `OpenOptions` by-value instead of by-ref, at the expense of a few extra copies. This can be done as the data is cheaply copyable, and we are in test code. But here, we are not, and while `File::try_clone` exists, it [issues system calls internally](https://github.com/rust-lang/rust/blob/1e746d7741d44551e9378daf13b8797322aa0b74/library/std/src/os/fd/owned.rs#L94-L111). Also, it would allocate an entirely new file descriptor, something that the fd cache was built to prevent.
* We change the `STORAGE_IO_TIME` metrics to support async.

Part of #4743.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

koivunej commented on pageserver/src/virtual_file.rs:
```diff
@@ -330,82 +402,11 @@ impl VirtualFile {
-            .observe_closure_duration(|| self.open_options.open(&self.path))?;
-
-        // Perform the requested operation on it
-        let result = STORAGE_IO_TIME
-            .with_label_values(&[op])
-            .observe_closure_duration(|| func(&file));
```
Why cannot this be in the macro as well? Maybe not with the observe_closure_duration, but with a timer for example?

arpad-m replied:
I dimly remember encountering metrics previously and it was seen as okay to remove them. If they are seen as useful here, i can obviously add a timer or use other means to measure the time.

arpad-m replied:
I've added back the metrics.

koivunej replied:
Not yet pushed.

koivunej replied:
Aah, I missed it.

koivunej commented on pageserver/src/virtual_file.rs:
```diff
@@ -208,6 +208,78 @@ impl CrashsafeOverwriteError {
+            .observe_closure_duration(|| $this.open_options.open(&$this.path))?;
+
+        // Perform the requested operation on it
+        let result = file.$($body)*?;
```
Shouldn't here be an await?

koivunej replied:
No, because we only do blocking actions.

koivunej replied:
Note the previous version did not use `?`... Which I am not totally sure if this uses, except that the `Ok(result)` in the bottom.

arpad-m replied:
Good catch about the `?`: I think it is not present because we want to put the file into the slot if there has been an error or not. It's pretty subtle, maybe I should put a comment.

arpad-m replied:
I'm now returning a lock object, check the code.

koivunej commented on pageserver/src/virtual_file.rs:
```diff
@@ -208,6 +208,78 @@ impl CrashsafeOverwriteError {
+/// opens it and evicts some other File if necessary. The passed parameter is
+/// assumed to be a function available for the physical `File`.
+///
+/// We are doing it via a macro as Rust doesn't support async closures that
+/// take on parameters with lifetimes.
```
I think async closures are not supported at all, but `FnOnce(&VirtualFile) -> impl std::future::Future<Output = Result<T, E>>` seems like it could work.

koivunej replied:
Aah oki, this wasn't about the body, which is invoked on a std::fs::File still. Where does the error come?

koivunej replied:
Maybe it's the Seek, which I didn't remember was also implemented for `&File`, weirdly enough.

Seems this is implementable as, so I am unsure why the macro:

```rust
pub(crate) async fn with_file<F, T>(&self, op: F) -> Result<T, std::io::Error>
  where F: FnOnce(&mut &std::fs::File) -> Result<T, std::io::Error> {}
```

koivunej replied:
Posted #5250 which seems to compile for me locally.

koivunej replied:
Aah ok, so maybe this is all for future future use?

koivunej replied:
Assuming this is just for future use, it still seems quite limited if you can only call methods on a file. Current plans have calling static fns with the file descriptor.

How about:

```
F: FnOnce(File) -> Fut,
Fut: std::future::Future<Output = (File, Result<T, std::io::Error>)>
```

Above would then work with:
- blocking calls to the file
- current async attempts
- maybe even tokio::fs?

arpad-m replied:
```rust
    pub(crate) async fn with_file<F, T>(&self, op: F) -> Result<T, std::io::Error>
    where
        F: FnOnce(&mut &std::fs::File) -> Result<T, std::io::Error>,
    {
```

#5250 is more or less the status quo: the functions we want to call will one day be async fn:


> Aah ok, so maybe this is all for future future use?

Yeah, that's what this PR is about (and all of my prior PRs for #4743 :) ). Quoting the issue's DOD:

> All code up to the point where we issue the system call to read from Image or Delta layer files is `async`.
The system call itself remains synchronous, it is _not_ converted to `tokio::fs`.
This allows merging the work in this PR without performance regressions caused by `tokio::fs`, thereby isolating any potential performance regressions to any refactoring changes required.

>  it still seems quite limited if you can only call methods on a file. Current plans have calling static fns with the file descriptor.

I don't know how else to pass the file to the snippet than via dot syntax. hygiene doesn't allow you to just access the `file` variable. I haven't tried out closures, as weird as it sounds they might work. It *might* be possible to change it to create a list of args or something, but what is possible in any case is to add a thin wrapper over `File` with a bunch of inherent functions.

koivunej replied:
Perhaps Heikki's proposal here is a better version? https://github.com/neondatabase/neon/pull/5247#issuecomment-1711200985

arpad-m replied:
Yeah guards might work and I think they would be nicer than a macro (if they work).

arpad-m replied:
> Where does the error come?

It's the same issue as in #5224 for the `test_files` function. See my comment there: https://github.com/neondatabase/neon/pull/5224#issuecomment-1710255989 . The error is quite analogous here:

```
error: lifetime may not live long enough
   --> pageserver/src/virtual_file.rs:516:39
    |
516 |             .with_file("read", |file| async move { file.read_at(buf, offset) })
    |                                 ----- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |                                 |   |
    |                                 |   return type of closure `[async block@pageserver/src/virtual_file.rs:516:39: 516:79]` contains a lifetime `'2`
    |                                 has type `&'1 std::fs::File`
```

If you have:

```Rust
    async fn with_file<F, R, FR>(&self, _op: &str, func: F) -> Result<R, Error>
    where
        F: FnOnce(&File) -> FR,
        FR: Future<Output = R>,
    {
```

See my branch [virtual_file_async_5_5](https://github.com/neondatabase/neon/compare/arpad/virtual_file_async_5_5) if you want to try it out yourself.

arpad-m replied:
>  I haven't tried out closures, as weird as it sounds they might work. 

They don't: :(

```
   --> pageserver/src/virtual_file.rs:520:53
    |
520 |         let result = with_file!(self, |file: &File| async { file.write_at(buf, offset) });
    |                                              -    - ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |                                              |    |
    |                                              |    return type of closure `[async block@pageserver/src/virtual_file.rs:520:53: 520:89]` contains a lifetime `'2`
    |                                              let's call the lifetime of this reference `'1`

``

koivunej replied:
I think this is good to close now because guard is used, though, together with the macro, but agreed, it does not work with `|file: &File|` we'd have to have `Arc<File>`.

koivunej replied:
Well, this is a better thread to continue than to start a new.

I still dislike the tt-matching macro for this simply because we cannot use for example tokio::fs::File or something else, so this will need to be resolved very soon after this PR..

Which I guess could be fine, but feels like we could find a way out on this PR.

Could we tt-munch a fake async closure? match `|$identifier:ident| async { ... }` then match that as `|$identifier| $body` then execute that as 

```rust
let res = {
  let $identifier = sl.as_ref();
  async { $body }.await;
};
```

(I don't know how to implement this, but I am assuming you do :)) It might actually be that the hygiene rules prohibit that.. If it were to work then we'd be able to express tokio::fs and and another way for this.

Though, I cannot see a way out the "other ways" from requiring the fd to be `Arc<OwnedFd|File>` because it needs to stay valid for the duration of the async operation which could be done on a userspace or kernel thread pool.

arpad-m replied:
I will try out your idea, it sounds good :+1:

hlinnaka commented:
> Lastly, we switch with_file to a macro. Mostly we do this because Rust has no support for async closures. One can make normal closures return a future, but that approach gets into lifetime issues the moment you want to pass data to these closures that has a lifetime. 

Another alternative to closures is to introduce a new guard object. So instead of:
```
self.with_file("foo", |file|.sync_all()).awat?
```
The callers would look like this:
```
{
    let file_guard = self.with_file();
    file_guard.sync_all().await?
    // `file_guard` is dropped
}
```
I guess in this case, the guard object would be just `slot_guard`, or a wrapper around it.

koivunej commented on pageserver/src/virtual_file.rs:
```diff
@@ -566,12 +559,19 @@ impl VirtualFile {
-        // unrelated I/O.
+        let handle = self.handle.get_mut();
+
+        // We don't have async drop so we cannot wait for the lock here.
+        // Instead, do a best-effort attempt at closing the underlying
+        // file descriptor by using `try_write`.
```
```suggestion
```

I think it's best to leave out speculation for likely not fitting features of the language.

arpad-m replied:
WDYM by not fitting? Why would async drop not fit here?

koivunej replied:
My personal opinion of how it probably is not a good fit for rust language :)

koivunej replied:
Let's just resolve this, fine to leave this.

arpad-m commented:
The idea with the guards was really great, I've moved most of the logic away from the macro into a function again that returns locks. There is still a macro to save repetition. I've also added back the metrics.

In the worst case, if it turns out that we really can't use the macro with async ops, we can always remove it again: now it's pretty cheap to do so (there is still some minor level of repetition that it saves).

hlinnaka commented on pageserver/src/virtual_file.rs:
```diff
@@ -566,12 +580,19 @@ impl VirtualFile {
+        // Instead, do a best-effort attempt at closing the underlying
+        // file descriptor by using `try_write`.
+        // This best-effort attempt should be quite good though
+        // as we have `&mut self` access. In other words, if the slot
+        // is still occupied by our file, we should be the only ones
+        // accessing it (and if it has been reassigned since, we don't
+        // need to bother with dropping anyways).
         let slot = &get_open_files().slots[handle.index];
-        let mut slot_guard = slot.inner.write().unwrap();
+        let Ok(mut slot_guard) = slot.inner.try_write() else { return };
+
```
Hmm, so if we don't get the lock, then the fd is kept open until it is evicted from the open files list. That's OK from a correctness point of view. I'm a little worried, though, that we might hold onto a large file that was deleted, so that the disk space cannot be reused.

hlinnaka replied:
Some alternatives:
- retry a few times, with a small delay. This is expected to be very rare, so that should be fine
- add a little global queue of dropped slots. Check that queue periodically and close all the underlying files
- just accept this as it is. Add an info! message for when it happens, so that we can keep an eye on it if it becomes a problem

I'm ok with any of those

arpad-m replied:
> retry a few times, with a small delay. This is expected to be very rare, so that should be fine

Personally I'd go with this.

As explained in the PR description, the only other component that can access the lock and cause the `try_lock` to fail is some other thread running the clock algorithm to get a free slot and is over that specific slot right in the moment where we drop the `VirtualFile`. This is unlikely, but gets more likely the more pressure is on the cache. In high-pressure situations however, leakages are being cleaned up quickly.

hlinnaka replied:
>For the rare else branch, I'd just spawn a task, do the write().await there.

Ah yes, that's a good idea, simple and robust

koivunej reviewed (changes requested):
Adding request changes for this thread at least: https://github.com/neondatabase/neon/pull/5247/files#r1321117325

arpad-m commented:
> I still have issues with the tt-muncher and lack of Arc (unless I missed it) but now the metrics are back, and this needs to be rebased.

regarding the tt muncher I have now implemented your suggestion to not use the dot syntax any more. For the lack of Arc, it's not really needed to use `Arc`s here, instead I used lock objects as suggested by Heikki (already done that at the time you made that comment). For the rebase, I've done it via a git merge.

koivunej reviewed (approved):
Ha, it worked! I'm okay with this now, even if the Arc changes might be still required. Also drop spawns a task, it should be good.
